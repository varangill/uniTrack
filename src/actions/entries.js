import uuid from 'uuid';
import database from '../firebase/firebase.js';

export const initAddEntry = (entryData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '', //these are default values for if nothing is passed in
      note = '',
      tag = "task",
      createdAt = 0
    } = entryData;
    const entryToAdd = {description, note, tag, createdAt}

    return database.ref(`users/${uid}/entries`).push(entryToAdd).then((ref) => { 
      dispatch(addEntry({
        id: ref.key,
        ...entryToAdd
      }));
    });
  };
};

// ADD_ENTRY
export const addEntry = (entry) => ({ 
  type: 'ADD_ENTRY',
  entry
});

// REMOVE_ENTRY
export const removeEntry = ({ id } = {}) => ({
  type: 'REMOVE_ENTRY',
  id
});

export const initRemoveEntry = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/entries/${id}`).remove().then(() => {
      dispatch(removeEntry({id}));
    });
  };
};

// EDIT_ENTRY
export const editEntry = (id, updates) => ({
  type: 'EDIT_ENTRY',
  id,
  updates
});

export const initEditEntry = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/entries/${id}`).update(updates).then(() => {
      dispatch(editEntry(id, updates));
    })
  }
};

// SET_ENTRIES
export const setEntries = (entries) => ({
  type: 'SET_ENTRIES',
  entries
})

export const startSetEntries = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/entries`).once('value').then((snapshot) => {
      const entries = [];

      snapshot.forEach((childSnapshot) => {
        entries.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });

      dispatch(setEntries(entries));
    });
  };
};