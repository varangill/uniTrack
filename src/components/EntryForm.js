import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


export default class EntryForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.entry ? props.entry.description : '',
      note: props.entry ? props.entry.note : '',
      tag: props.entry ? props.entry.tag : 'task',
      createdAt: props.entry ? moment(props.entry.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onTagChange = (e) => {
    const tag = e.target.value;
    this.setState(() => ({ tag }));
    console.log("tag")
  }
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description) {
      this.setState(() => ({ error: 'Please provide a title.' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        tag: this.state.tag,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          type="text"
          className="input-items__text"
          placeholder="Description"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <select 
          value={this.state.tag}
          onChange={this.onTagChange}
          className="input-items__select"
        >
          <option value="task">Task</option>
          <option value="goal">Goal</option>
        </select>
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          placeholder="Add a note for your entry (optional)"
          className="input-items__textarea"
          value={this.state.note}
          onChange={this.onNoteChange}
        >
        </textarea>
        <div>
          <button className="button">Save Entry</button>
        </div>
      </form>
    )
  }
}
