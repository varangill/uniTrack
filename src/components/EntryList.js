import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import EntryListItem from './EntryListItem';
import selectEntries from '../selectors/entries';

export const EntryList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-mobile">Entries</div>
      <div className="show-desktop">Entry</div>
      <div className="show-desktop">Type</div>
    </div>
    <div className="list-main">
      {
        props.entries.length === 0 ? (
          <div className="list-entry list-entry--msg">
            <span>No Entries</span>
          </div>
        ) : (
            props.entries.map((entry) => {
              return <EntryListItem key={entry.id} {...entry} />; //removed return
            })
          )
      }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    entries: selectEntries(state.entries, state.filters)
  };
};

export default connect(mapStateToProps)(EntryList);
