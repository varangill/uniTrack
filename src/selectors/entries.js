import moment from 'moment';

// Get visible entries

export default (entries, { text, sortBy, startDate, endDate, tag }) => {
  return entries.filter((entry) => {
    const createdAtMoment = moment(entry.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = entry.description.toLowerCase().includes(text.toLowerCase());
    const tagMatch = tag === 'all' ? true : tag === entry.tag;
    
    return startDateMatch && endDateMatch && textMatch && tagMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'alphabet') { //this is alphabet now
      return a.description.localeCompare(b.description);
    }
  });
};
