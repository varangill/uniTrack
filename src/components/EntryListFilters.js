import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAlphabet, setStartDate, setEndDate, setTypeFilter } from '../actions/filters';
import { Link } from 'react-router-dom';

export class EntryListFilters extends React.Component {
  state = {
    calendarFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  }

  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'alphabet') {
      this.props.sortByAlphabet();
    }
  }

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };

  onTypeChange = (e) => {
    this.props.setTypeFilter(e.target.value);
  }

  render() {
    return (
      
      <div className="content-container">
        <div className="subheader__actions">
          <center>
          <Link to="/create" className="button">Create Entry</Link>
          </center>
        </div>

        <div className="input-items">
          <div className="input-items__item">
            <input
              type="text"
              placeholder="Search entries..."
              className="input-items__text"
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-items__item">
            <select
              className="input-items__select"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="alphabet">Alphabet</option>
            </select>
          </div>
          <div className="input-items__item">
            <select 
              className="input-items__select"
              onChange={this.onTypeChange}
            >
              <option value="all">All</option>
              <option value="task">Task</option>
              <option value="goal">Goal</option>
            </select>
          </div>
          <div className="input-items__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  setTypeFilter: (type) => dispatch(setTypeFilter(type)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAlphabet: () => dispatch(sortByAlphabet()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(EntryListFilters);
