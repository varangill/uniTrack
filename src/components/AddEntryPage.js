import React from 'react';
import { connect } from 'react-redux';
import EntryForm from './EntryForm';
import { initAddEntry } from '../actions/entries';

export class AddEntryPage extends React.Component {
  onSubmit = (entry) => {
    this.props.initAddEntry(entry);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="subheader">
          <div className="content-container">
            <h1 className="subheader__title">Add Entry</h1>
          </div>
        </div>
        <div className="content-container">
          <EntryForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  initAddEntry: (entry) => dispatch(initAddEntry(entry))
});

export default connect(undefined, mapDispatchToProps)(AddEntryPage);
