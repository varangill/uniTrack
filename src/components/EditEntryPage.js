import React from 'react';
import { connect } from 'react-redux';
import EntryForm from './EntryForm';
import { initEditEntry, initRemoveEntry, initAddEntry } from '../actions/entries';

export class EditEntryPage extends React.Component {
  

  onSubmit = (entry) => {
    this.props.initEditEntry(this.props.entry.id, entry);
    this.props.history.push('/');
  };

  onDelete = () => {
    this.props.initRemoveEntry({ id: this.props.entry.id });
    this.props.history.push('/');
  };

  onDuplicate = () => {
    this.props.initAddEntry(this.props.entry);
    this.props.history.push('/'); //if possible, push user to new duped entry
  };

  render() {
    return ( //if possible, somehow make duplicate and remove buttons on same line as add button by puttig them in EntryForm
      <div>
        <div className="subheader">
          <div className="content-container">
            <h1 className="subheader__title">Edit Entry</h1>
          </div>
        </div>
        <div className="content-container">
          <EntryForm
            entry={this.props.entry}
            onSubmit={this.onSubmit}
          /> 
          <button className="button button--duplicate" onClick={this.onDuplicate}>Duplicate</button> 
          <br /> 
          <button className="button button--remove" onClick={this.onDelete}>Remove</button>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state, props) => {
  return {
    entry: state.entries.find((entry) => entry.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  initEditEntry: (id, entry) => dispatch(initEditEntry(id, entry)),
  initRemoveEntry: (data) => dispatch(initRemoveEntry(data)),
  initAddEntry: (entry) => dispatch(initAddEntry(entry)) //this duplicates
});

export default connect(mapStateToProps, mapDispatchToProps)(EditEntryPage);
