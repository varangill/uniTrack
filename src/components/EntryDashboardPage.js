import React from 'react';
import EntryList from './EntryList';
import EntryListFilters from './EntryListFilters';

const EntryDashboardPage = () => (
  <div>
    <EntryListFilters />
    <EntryList />
  </div>
);

export default EntryDashboardPage;
