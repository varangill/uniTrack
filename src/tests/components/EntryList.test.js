import React from 'react';
import { shallow } from 'enzyme';
import { EntryList } from '../../components/EntryList';
import entries from '../fixtures/entries';

test('should render EntryList with entries', () => {
  const wrapper = shallow(<EntryList entries={entries} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render EntryList with empty message', () => {
  const wrapper = shallow(<EntryList entries={[]} />);
  expect(wrapper).toMatchSnapshot();
});
