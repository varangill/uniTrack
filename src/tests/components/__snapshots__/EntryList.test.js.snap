// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EntryList with empty message 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-mobile"
    >
      Entries
    </div>
    <div
      className="show-desktop"
    >
      Entry
    </div>
    <div
      className="show-desktop"
    >
      Type
    </div>
  </div>
  <div
    className="list-main"
  >
    <div
      className="list-item list-item--msg"
    >
      <span>
        No Entries
      </span>
    </div>
  </div>
</div>
`;

exports[`should render EntryList with entries 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-mobile"
    >
      Entries
    </div>
    <div
      className="show-desktop"
    >
      Entry
    </div>
    <div
      className="show-desktop"
    >
      Type
    </div>
  </div>
  <div
    className="list-main"
  >
    <EntryListItem
      createdAt={0}
      description="Gum"
      id="1"
      key="1"
      note=""
      tag="task"
    />
    <EntryListItem
      createdAt={-345600000}
      description="Rent"
      id="2"
      key="2"
      note=""
      tag="task"
    />
    <EntryListItem
      createdAt={345600000}
      description="Credit Card"
      id="3"
      key="3"
      note=""
      tag="task"
    />
  </div>
</div>
`;
